### ISSUE TITLE
Segmentation with Myanmar tonemarks 1069–106D

### ISSUE THREAD

**Issue by ohbendy** [harfbuzz/harfbuzz#4974]
I'm not sure if this is a Harfbuzz bug or something else so please let me know where to raise the issue if it's not here :) 

I notice some bug in segmentation where tonemarks 1069–106D get separated from their syllable and appear after a linebreak in Firefox and Chrome. This text is Asho language but the tonemarks also occur in Western Pwo.

![Tonemarks stranded](https://github.com/user-attachments/assets/0a452274-eee8-40ad-8cb1-175501235eb7)

Also, Latin punctuation signs (comma, period, parentheses) don't stick to these tonemarks and get pushed onto a new line:

![Punctuation stranded](https://github.com/user-attachments/assets/2ad41768-1972-41a9-9ea4-8904306fc01c)

**Comment by jfkthame** [comment]
I don't think this is a harfbuzz issue; it's an issue with the line-breaking algorithms/libraries used by the browser or other application.

In the case of Firefox, if you set the `intl.icu4x.segmenter.enabled` pref to `false` in `about:config`, these breaks will no longer happen (at least for the example I tested), but on the other hand it will generally do a poor job of line-breaking as it will not find many perfectly reasonable breaks.

You could file bugs against the browsers, but they in turn will probably direct you to upstream projects like ICU4C and ICU4X. In the case of Firefox, at least, the ICU4X Segmenter is the relevant project; I'm not sure what Chrome is currently using. 

https://github.com/unicode-org/icu4x
https://unicode-org.atlassian.net/projects/ICU

**Comment by behdad** [comment]
cc @drott

**Comment by drott** [comment]
@ohbendy could you provide a link to the test content that you used for the screenshots?

**Comment by drott** [comment]
Filed as https://crbug.com384453959 - if you have test content, it would also be helpful if you could link to it, or attach it to the Chrome issue.

**Comment by ohbendy** [comment]
Here's an Asho paragraph which includes the Latin punctuation:

 မွ့ၩထၩစိၪတကွ့ၪအဆဳၭကူၩ့ငှိၩဇီၭ ဆမံၩဟဲၪဧၩဧစိၩဘၩဧဴၩဖုၩ ဆံၩ့နှိၩပဳၩစိၩအူၭနၨၩ ဆမံၩနုၪဗိၪ, ထဲးၩပွံၩပၩဘံၩ ဟဲၪဖုၩထီါၩ့ဧ့ၩဧၧၭ. နံၩကူၩ့ဆဳၭဆမံၩဟဲၪနၨၩ လူၭစိၪလွံၩဆီါၬဝဲးၩဧၧၭ, ခွံၪ(ပူၩ့) ည့ၪဝဲးၩဧၧၭ, စိၩအူၭလွံၩဟုၪတုၩ, ဧဘံၩ့တုၩ, အလူၩ့နဳၬဝဲးၩဧၧၭ. နံၩကူၩ့ဆဳၭခၠၨၩ့ဧၪစိၩလူၩ့(စိၩလုၪ့) အတွးၭ အခၠးၭ(ဂၬနၪ့)ဟဲၪ ဇိၬကဴၩဧလုၩ ဟီၪ့လၭ. ဆီါၬဝဲးၩမျုၩ့, ယဳၬဝဲးၩမျုၩ့, ည့ၪဝဲးၩမျုၩ့ဒၧၩမွ့ၩဟီၪ့ဧၧၭ. အဆဳၭ ကူၩ့ငှိၩ, ကူၩ့ဆိၭဆမံၩဟဲၪဧၩလုၩအိၭ ခၠဲးၭနၧၬစိၩဇိၬပဲးၬနၭ ဆဳၭအီၭစိၩအကၠံၩဆံၩဧၧၭ. နံၩခၠၨၩ့ဧၪ ဆမံၩဖုၩဘဳၩ့လုၩ ဧၧၭမၧၭနိၪနၩ, ဆမံၩဧၪမအီါၩ့ပိၩနဳၭလၩဧဴၩ. ပီၪ့နှဲၬပီၪ့နှဲၬ မအီါၩ့ပိၩနဳၭလၩဧဴၩ. စိၩအူၭဒုၭကၪ မမှုၩလၩဂုၬ လူၭစိၪဟဲၪ ဘဳၩ့စဴၪဧၧၭ, အဘဳၩ့ဟဳၩ့ဧၧၭ, ဘဳၩ့ခၭဆွ့ၩဆၪအီါၪ့ဧၧၭ, ဘဳၩ့ခၭတံၪ အၨၭကၧၭ, ဘဳၩ့ခၭတံၪထၧၪဧၧၭနဳၭဂုၬဟဲၪမှၭဝဲးၩ, ဆီါၬဝဲးၩ ဟံၩဇၩဝဲးၩဧၧၭ. နံၩလှုၪဝဲးၩမျိၩ့မွ့ၩအိၩ့မျှိၭ တၩမပဲးၭလၩဧဴၩ. အလၧၩနဳၭ, အဂျီၩ့နဳၭဧ့ၩဝဲးၩမျုၩ့ပဳၪ လိၪအိၩ့ မစဴၪပိၩနဳၩလၩဧဴၩ. ဖိၪမ့ၪ(ယၬမ့ၪ)မျုၩ့တၩဆၪဧ့ၩထ့ၩအိၩ့, လူၭအီါၪ့ ထ့ၩအိၩ့, ဟဳၩထ့ၩအိၩ့, စိၬ ထ့ၩအိၩ့ဘုၭမပဲးၭလၩဧဴၩ. မျုၩ့ဆဳၭဧဆၩဧ့ၩဗဴၩဆ့ၭ, မျှိၭတၩဘွ့ၩ. ခၪဧဴၩဓံၩ, ငီၭကဴၩဓံၩ. အီါၩ့ဟီါၪ့ပွံၩ နထီါၩ့ ဧ့ၩလၩဧဴၩ. စိၩအူၭဒုၭကၪခီါၪ့ဂုၬဇူၪ့မှၭထ့ၩအိၩ့ယဳၬ လှုၩပဳၩမဟဳၩ လၩဧဴၩ. စိၩအူၭဒုၭကၪဂုၬမွှဲးၪပွဲးၩ, မွှဲးၪဆၨၪမှုၪထ့ၩအိၩ့, ခွံၪ, စူၭနဳၬဒူၪ့ဟဲၪ, အမျုၩ့ဆဳၭဇူ မှၭလူၭဆီါၪ့ ထ့ၩအိၩ့, မအီါၩဧ့ၩနဳၭလၩဧဴၩ. အိၩမပဲးၭလၩဧဴၩ. ဆမံၩဟဲၪစိၩအူၭလွံၩဧၧၪအီါၪ့ၭ, လူၩ့ နဳၬ, ဧဓုၩနှဴၩဧထီၪ့မျှီၪ့ဗဴၩဆ့ၭ. "စဴၪထ့ၩစဘဳၩ့, ဟဳၩထ့ၩစဘဳၩ့" ဓံၬဟဳၩနၧၬမအီါၩဧ့ၩနဳၭလၩဧဴၩ.
 
 (Note there's also an issue with Microsoft's syllable model which I've raised here: https://github.com/MicrosoftDocs/typography-issues/issues/1187)

**Comment by ohbendy** [comment]
I just also remembered I was [testing breakpoints](https://github.com/ohbendy/Myanmar-font-resources/blob/master/Segmentation.md) a while ago, @drott you might find the issue above is part of a broader problem.

**Comment by kojiishi** [comment]
I can't reproduce in a simple HTML/CSS with the text from the above comment.
https://output.jsbin.com/qakomar

Does the page have any CSS properties? Full HTML to reproduce is appreciated.

**Comment by ohbendy** [comment]
That page shows the problem when I view it on Firefox using the default Noto Sans Myanmar:

![Image](https://github.com/user-attachments/assets/c14e26be-ecbb-4006-8598-8361c36e93e4)

And also when using our WIP font where I'm suppressing dotted circles:

![Image](https://github.com/user-attachments/assets/d6641d92-07f6-41a5-9233-c72ad1ce88e5)

**Comment by jfkthame** [comment]
What version of Firefox? I can reproduce this (as I understand it) with Firefox 134, but not with a Nightly build (Firefox 136), suggesting that a recent change may have fixed it. I suspect the update to harfbuzz 10.2.0 in https://bugzilla.mozilla.org/show_bug.cgi?id=1941147 is probably the relevant fix.

I haven't investigated this, but a possibly-relevant change that was part of that update may be https://github.com/harfbuzz/harfbuzz/commit/f887096ab11239def01fcee44cf94cd4b34ed0ab.
